vivisect/tests/testswitches.py

```diff
@@ -0,0 +1,187 @@
+
+
+def genMsSwitchWorkspace(maps=MS_MAPS, mapbase=0x400000, bufferpages=2):
+    vw = vivisect.VivWorkspace()
+    vw.setMeta('Architecture', 'i386')
+    vw.setMeta('Platform', 'windows')
+    vw.setMeta('Format', 'pe')
+    vw._snapInAnalysisModules()
+
+    applyMaps(vw, maps, mapbase, bufferpages)
+    return vw
+
+def genGccSwitchWorkspace(maps=LIBC_MAPS, mapbase=0x400000, bufferpages=2):
+    vw = vivisect.VivWorkspace()
+    vw.setMeta('Architecture', 'i386')
+    vw.setMeta('Platform', 'linux')
+    vw.setMeta('Format', 'elf')
+    vw._snapInAnalysisModules()
+
+    applyMaps(vw, maps, mapbase, bufferpages)
+    return vw
+
+cbs_ms_0 = [  
+        (4194304, 13, 4194304),
+        (4194317, 9, 4194304),
+        (4194326, 9, 4194304),
+        (4194335, 38, 4194304),
+        (4194373, 14, 4194304),
+        (4194387, 10, 4194304),
+        (4244306, 12, 4194304),
+        (4244318, 27, 4194304),
+        (4274127, 6, 4194304),
+        (4274133, 9, 4194304),
+        (4274142, 9, 4194304),
+        (4274151, 5, 4194304),
+        (4274156, 14, 4194304),
+        (4308616, 17, 4194304),
+        (4308633, 26, 4194304),
+        (4308659, 26, 4194304),
+        (4308685, 26, 4194304),
+        (4308711, 15, 4194304)]
+
+class MsSwitchTest(unittest.TestCase):
+    def test_ms_switch_0(self):
+        vw = genMsSwitchWorkspace(MS_MAPS, 0x400000)
+        vw.makeFunction(0x400000)
+        
+        self.assertEqual(vw.getFunctionBlocks(0x400000), cbs_ms_0)
+        #self.assertEqual(vw.getXrefsFrom(0x
+
+
+
+
+
+class LibCSwitchTest(unittest.TestCase):
+    def test_libc_switch_0(self):
+        vw = genGccSwitchWorkspace(LIBC_MAPS, 0x500000)
+
+
+#=======  test generator code =======
+
+import visgraph.pathcore as vg_path
+import vivisect.impemu.monitor as viv_imp_monitor
+
+MAX_GAP = 100
+MAX_INSTR_SIZE = 20
+
+class AnalysisMonitor(viv_imp_monitor.AnalysisMonitor):
+    def __init__(self, vw, fva):
+        viv_imp_monitor.AnalysisMonitor.__init__(self, vw, fva)
+        self.vas = []
+        
+    def prehook(self, emu, op, starteip):
+        #print "0x%x: %r" % (op.va, op)
+        if starteip not in self.vas:
+            self.vas.append(starteip)
+
+def analyzeFunction(vw, fva):
+    maps = getFuncMaps(vw, fva)
+    return maps
+
+
+def getFuncMaps(vw, fva):
+    opvas, refvas = getVas(vw, fva)
+
+    vas = list(opvas)
+    vas.extend(refvas['reads'])
+    vas.sort()
+
+    maps = []
+    lastva = startva = vas[0]
+    for va in vas:
+        #print hex(va), hex(lastva), hex(startva)
+        if (va - lastva) > MAX_GAP:
+            size = lastva+MAX_INSTR_SIZE - startva
+            memory = vw.readMemory(startva, size)
+            maps.append((startva, size, memory))
+            startva = va
+
+        lastva = va
+
+    # grab the last map
+    size = lastva+MAX_INSTR_SIZE - startva
+    memory = vw.readMemory(startva, size)
+    maps.append((startva, size, memory))
+
+    return maps
+
+def getVas(vw, fva):
+    global emu
```

Comment:
sooo...does this work? Because emu isn't defined at this point.

Reply:
it does.  but you got me to find another bug.  thanks!